### ISSUE TITLE
Guidance for constructing a quadratic function from the solvers FunctionInfo

### ISSUE THREAD

**Issue by johannahaffner** [patrick-kidger/optimistix#107]
Hi Patrick, 

I'm implementing a descent that solves an iterative quadratic problem for sequential quadratic programs.
Currently, I'm swapping out the `NewtonDescent` in `BFGS` for a descent that calls `optx.quadratic_solve` in its `query`, rather than `lx.linear_solve` (since we don't know which inequality constraints are active). 

This means that at every step, BFGS approximates the Hessian and gradient, which I then use to construct a problem

$\min q(x) = 1/2 x^T G x + c^T x \qquad s.t. \; \; Ax \geq b$ 

and iteratively solve it for $x$, which is then used to define the (unscaled) step. 

For this to be efficient, I obviously don't want to re-compute the Hessian approximation, since we already have that (quadratic solvers also only do that once, in their `init`). So I wrote a method that exports `FunctionInfo.EvalGradHessian` to a `QuadraticApproximation`, with an attribute `.quadratic`. I do an instance check in the `init` of the quadratic function, but that one does not evaluate to `True`, because we have other things that happen before, like transformation to `NoneAux`, `OutAsArray`, and so on. (I.e. `isinstance(fn, QuadraticApproximation)` is always `False`.

Which, finally, brings me to the question: how do I sneak in previously computed second-order information?  Some tag? 
What I want is _kind of_ a warm start, which could be an alternative but seems much trickier to implement.

(Also: things are finally coming together a bit, let's talk code soon.)

**Comment by patrick-kidger** [comment]
Can you put it in the solver state? See for example how `Chord` keeps around its initial evaluation of the Jacobian, as compared to a Newton solve:

https://github.com/patrick-kidger/optimistix/blob/dcafc480d4a2bb9ec3b62658d5ae136db847b2d8/optimistix/_solver/newton_chord.py#L124

**Comment by johannahaffner** [comment]
Ok, maybe I have to be more precise. Normally we would have

```python
class Solver(...):
    ...
    search
    descent
```
where descent has a linear solver as an attribute.

Now in some solvers, we actually use another optimistix solver internally - for example in `IndirectDampedNewtonDescent`, we have a root finder.
I want to write a descent that solves a quadratic subproblem, using a quadratic solver. This quadratic solver comprises a search and descent as well, and the linear solve then takes place in the descent of the quadratic solver.
Does that make sense? So there is a nested structure

- Outer solver: solves f(y, args) s.t. c(y) >= 0
    - outer search
    - outer descent
         - inner solver: solves q(x), st. A(x) >= b
              - inner search
              - inner descent: appropriate constrained descent, e.g. interior-point
    
So the inner solver treats the quadratic approximation as the target function, and works with linearised constraints.
Now, in the "outer descent", I just want to call `optx.quadratic_solve(q(x), quadratic_solver, y0, ...)`. The thing is - I know the Hessian of `q` already: it is the Hessian approximation made by the outer solver, and in principle I already have a lineax linear operator! 
Right now I'm computing the Hessian of `q` in the quadratic solver's `init`, and then I do indeed keep it around. But I'd like to avoid computing it in the first place, it just seems redundant. Since I already have it, I'd like it to go straight into the solver state.

Does that make sense? By the way, my reason for solving a whole quadratic subproblem in the outer descent rather than just using an interior descent as the outer descent, e.g. in `BFGS`, is that I think this will be more economical and converge better. In the literature this is known as "iterative refinement". You can think of this as actually extracting all the information we have in the quadratic approximation, before we go to the trouble of computing another Hessian approximation.

Something I could imagine is creating a method `warm_init` for the quadratic solver, which directly takes a`FunctionInfo.EvalGradHessian`, and extracts what it needs to construct its own state. Similar to how it's done in lineax, we could then call `optx.quadratic_solve(..., state=state)` and bypass the pre-processing steps done in the top-level APIs. But then FunctionInfo becomes sort-of-public, even if we keep everything undocumented.

**Comment by patrick-kidger** [comment]
IIUC you currently have a symmetric `lx.AbstractLinearOperator`, and then you are defining
```python
def q(x):
    return jnp.dot(x, 0.5 * op.mv(x) + c)
```
?

Hmm. So I think if you then define `op2 = lx.JacobianLinearOperator(jax.grad(q), x)` every time you need the Hessian (don't even save it), then I think you'd get very similar behaviour (for both `.mv` and `.as_matrix`) as just somehow smuggling in `op` directly!

I think it would probably be slightly inefficient because the above doesn't exploit the assumed symmetry of `op`, so the gradients would flow through both of the `x`s in my definition of `q` above, and we'd end up computing `0.5 * (op + op.T)`. But probably that is fixable like so:
```python
@jax.custom_jvp
def q(x):
    return jnp.dot(x, 0.5 * op.mv(x) + c)

@q.defjvp
def q_jvp(p, t):
    (x,) = p
    (tx,) = t
    return jnp.dot(tx, op.mv(x) + c)
```

One possible concern is if the eventual linear solver materialises the matrix, and we don't want to do that in every step: in this case the *original* `op` (the one used to define `q`, before we do any of this) should be replaced with `op = lx.materialise(op)`.

Taking a step back, wrapping JAX functions in a `jax.custom_jvp` is arguably the appropriate way of passing in higher-order information to *any* API. (Without that API -- in this case `optx.quadratic_solve` -- ever needing to consume it as a separate argument.)

**Comment by johannahaffner** [comment]
For every step in the inner solver, I need access both to `op` and to `op.mv(x) + c`. 
Right now, in the inner (quadratic) solver `init` I use `op2 = lx.PyTreeLinearOperator(jax.hessian(q)(y))`, and update the gradient in `step`, I think already with custom jvp such as the one you are suggesting.
`op2` is then one of the operators used in the definition of the KKT system, where I call `op2.mv`. 

Since $q$ is a quadratic, it's Hessian is a constant, and after taking the Hessian and converting it to a linear operator again, we get `op2.as_matrix() == op.as_matrix()`. Since `op` is just a linear operator, it's Hessian might (a lot) cheaper to compute than the one of the target function, though?

Thanks for thinking of the materialisation thing! I'll add that just in case.

**Comment by johannahaffner** [comment]
Closing this for now, we can revisit it if we see that it is a performance bottleneck.
